README.md
# AutoFish Forge Mod by FreneticFeline

AutoFish Forge Mod is a Minecraft mod that provides the option to automatically reel
in and re-cast the fishing rod when a nibble is detected.  Primary ptions include
MultiRod, which automatically switches to the next available fishing rod in the hotbar
when the current rod breaks; and Break Protection, which will stop using a rod before
it breaks.  Additional options are available.

AutoFish bite detection is 100% accurate in Single Player games, where the mod has access to the
world server's state.  In MultiPlayer games, the mod must rely on the same hints given to players:
bobbing of the fish hook, water wake approaching the fish hook, the splash sounds, etc.  Therefore,
in MultiPlayer, bite detection is less accurate (around 85%).  See the "Aggressive Bite Detection"
option.

## Credits
This mod was originally a blatant copy of the functionality provided by the LiteMod
Autofish mod by troyboy50.  It is not affiliated with or endorsed by troyboy50.

## Installation Requirements
AutoFish Forge Mod versions 1.x-y.z require Minecraft 1.x with compatible version
of MinecraftForge installed.

## Installation Instructions
Copy the `mod_autofish_forge-1.x-y.z.jar` file to the `mods` directory in your Minecraft
data directory.

## Usage Instructions
By default, AutoFish is enabled, but all other options are disabled.  To change
these options, Click the **[Mods]** button from the opening Minecraft screen, locate the
AutoFish Forge Mod in the mods list, and click the **[Configure]** button.

The AutoFish Forge Mod options can also be reached in-game using the configured hotkey.
_Note that the hotkey will only function if the item currently in use by the player is a fishing rod._
By default, the hotkey is the letter `o`, but this binding can be changed or disabled in the
standard Minecraft **Controls** screen.

## Configuration Options

### Enable AutoFish
Automatically reel in and re-cast when a fish nibbles the hook. If set to false, all other AutoFish
functionality is disabled.

### Enable MultiRod
Automatically switch to a new fishing rod when the current rod breaks, if one is available in the hotbar.

### Enable Break Protection
Stop fishing or switch to a new rod just before the current rod breaks.  Useful if you want to repair
your enchanted fishing rods.

### Re-Cast Delay
Time (in seconds) to wait before automatically re-casting. Increase this value if server lag causes
re-casting to fail.

### Enable Entity Clear Protection
When playing on a server, re-cast after the server clears entities.  Useful if you are playing on a
server that periodically deletes all entities (including fishing hooks) from the world, which causes
you to stop fishing.

### Enable Aggressive Bite Detection
When playing on a server, be more aggressive about detecting fish bites.  Improves multiplayer bite
detection from ~85% to ~95%, but false positives will be more likely, especially if other players are
fishing very close by.

### Handle Problems
[HAS SIDE EFFECTS] Re-cast when problems detected (e.g. if not in water or if a MOB has been hooked).
If enabled, non-fishing use of the fishing rod (e.g. intentionally hooking MOBs) will be affected.

### Enable Fast Fishing
[SINGLE PLAYER ONLY] Fish will bite right after casting.


## Development Setup Instructions
Follow the standard Forge mod development setup instructions.  They can be found
in the file named `README-MinecraftForge.txt`


CHANGELOG.txt
1.12-1.7
========
Slightly improved bite detection for multiplayer with option to be even more aggressive.

1.12-1.6
========
Fixed bug: Multirod switching sometimes fails when Break Protection is off
Fixed bug (hopefully): Recast sometimes fails

1.12-1.5
========
Added Fast Fishing option for Single Player
Mod compatibility improvements (Thanks BoxOfFlex)

1.12-1.4
========
Added configuration option for recast delay

1.12-1.3
========
Update to work on Minecraft/Forge 1.12
Added experimental Entity Clear protection

1.11-1.2
========
Update to work on Minecraft/Forge 1.11.2

1.10-1.2
========
Update to work on Minecraft/Forge 1.10 and 1.10.2

1.9.4-1.2
=======
Update to work on Minecraft/Forge 1.9.4

1.9-1.2
=======
Update to work on Minecraft/Forge 1.9 (Thanks Maaiins)
More reliable technique for detecting fish bite in single player

1.8.9-1.1
=======
Update to work on Minecraft 1.8.9. (Thanks SpacyNG)

1.8-1.1
=======
Implement a hotkey to reach the AutoFish options screen while in-game.

1.8-1.0
=======
Initial release of the AutoFish Forge Mod


src/main/java/net/unladenswallow/minecraft/autofish/ModAutoFish.java
package net.unladenswallow.minecraft.autofish;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import net.minecraftforge.common.config.Configuration;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.SidedProxy;
import net.minecraftforge.fml.common.Mod.EventHandler;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import net.unladenswallow.minecraft.autofish.events.EventListener;
import net.unladenswallow.minecraft.autofish.proxy.CommonProxy;
import net.unladenswallow.minecraft.autofish.util.Logger;

@Mod(modid = ModAutoFish.MODID, useMetadata = true, acceptedMinecraftVersions="[1.8.9,1.9)", acceptableRemoteVersions="[1.8.9,1.9)",
guiFactory = "net.unladenswallow.minecraft.autofish.gui.GuiFactory")
public class ModAutoFish {
    public static final String MODID = "mod_autofish";
    
    public static Configuration configFile;
    public static boolean config_autofish_enable;
    public static final boolean CONFIG_DEFAULT_AUTOFISH_ENABLE = true;
    public static boolean config_autofish_multirod;
    public static final boolean CONFIG_DEFAULT_AUTOFISH_MULTIROD = false;
    public static boolean config_autofish_preventBreak;
    public static final boolean CONFIG_DEFAULT_AUTOFISH_PREVENTBREAK = false;
    public static boolean config_autofish_entityClearProtect;
    public static final boolean CONFIG_DEFAULT_AUTOFISH_ENTITYCLEARPROTECT = false;
    public static int config_autofish_recastDelay;
    public static final int CONFIG_DEFAULT_AUTOFISH_RECASTDELAY = 2;
    public static boolean config_autofish_fastFishing;
    public static final boolean CONFIG_DEFAULT_AUTOFISH_FASTFISHING = false;
    public static boolean config_autofish_aggressiveBiteDetection;
    public static final boolean CONFIG_DEFAULT_AUTOFISH_AGGRESSIVEBITEDETECTION = false;
    public static boolean config_autofish_handleProblems;
    public static final boolean CONFIG_DEFAULT_AUTOFISH_HANDLEPROBLEMS = false;
    
    private static final List<String> propertyOrder = new ArrayList<String>(Arrays.asList(new String[] {
            "Enable AutoFish",
            "Enable MultiRod",
            "Enable Break Protection",
            "Re-Cast Delay",
            "Enable Entity Clear Protection",
            "Enable Aggressive Bite Detection",
            "Handle Problems",
            "Enable Fast Fishing"
    }));
    
    @SidedProxy(clientSide="net.unladenswallow.minecraft.autofish.proxy.ClientProxy", serverSide="net.unladenswallow.minecraft.autofish.proxy.ServerProxy")
    public static CommonProxy proxy;
    
    public static AutoFish autoFish = new AutoFish();
    public static EventListener eventListener = new EventListener(autoFish);
    
    public static ModAutoFish instance = new ModAutoFish();
            
    @EventHandler
    public void preInit(FMLPreInitializationEvent preInitEvent) {
        ModAutoFish.proxy.preInit(preInitEvent);
        configFile = new Configuration(preInitEvent.getSuggestedConfigurationFile());
        configFile.renameProperty(Configuration.CATEGORY_GENERAL, "Fast Fishing", "Enable Fast Fishing");
        configFile.renameProperty(Configuration.CATEGORY_GENERAL, "Enable Enity Clear Protection", "Enable Entity Clear Protection");
        configFile.setCategoryPropertyOrder(Configuration.CATEGORY_GENERAL, propertyOrder);
        syncConfig();
    }
    
    @EventHandler
    public void init (FMLInitializationEvent event) {
        Logger.info("Initializing " + ModAutoFish.MODID);
        ModAutoFish.proxy.init(event);
    }

    public static void syncConfig() {
        config_autofish_enable = configFile.getBoolean("Enable AutoFish", Configuration.CATEGORY_GENERAL, CONFIG_DEFAULT_AUTOFISH_ENABLE, "Automatically reel in and re-cast when a fish nibbles the hook. If set to false, all other AutoFish functionality is disabled.");
        config_autofish_multirod = configFile.getBoolean("Enable MultiRod", Configuration.CATEGORY_GENERAL, CONFIG_DEFAULT_AUTOFISH_MULTIROD, "Automatically switch to a new fishing rod when the current rod breaks, if one is available in the hotbar.");
        config_autofish_preventBreak = configFile.getBoolean("Enable Break Protection", Configuration.CATEGORY_GENERAL, CONFIG_DEFAULT_AUTOFISH_PREVENTBREAK, "Stop fishing or switch to a new rod just before the current rod breaks.  Useful if you want to repair your enchanted fishing rods.");
        config_autofish_entityClearProtect = configFile.getBoolean("Enable Entity Clear Protection", Configuration.CATEGORY_GENERAL, CONFIG_DEFAULT_AUTOFISH_ENTITYCLEARPROTECT, "When playing on a server, re-cast after the server clears entities.  Useful if you are playing on a server that periodically deletes all entities (including fishing hooks) from the world, which causes you to stop fishing.");
        config_autofish_recastDelay = configFile.getInt("Re-Cast Delay", Configuration.CATEGORY_GENERAL, CONFIG_DEFAULT_AUTOFISH_RECASTDELAY, 1, 10, "Time (in seconds) to wait before automatically re-casting. Increase this value if server lag causes re-casting to fail.");
        config_autofish_fastFishing = configFile.getBoolean("Enable Fast Fishing", Configuration.CATEGORY_GENERAL, CONFIG_DEFAULT_AUTOFISH_FASTFISHING, "[SINGLE PLAYER ONLY] Fish will bite right after casting.");
        config_autofish_aggressiveBiteDetection = configFile.getBoolean("Enable Aggressive Bite Detection", Configuration.CATEGORY_GENERAL, CONFIG_DEFAULT_AUTOFISH_AGGRESSIVEBITEDETECTION, "When playing on a server, be more aggressive about detecting fish bites.  Improves multiplayer bite detection from ~85% to ~95%, but false positives will be more likely, especially if other players are fishing very close by.");
        config_autofish_handleProblems = configFile.getBoolean("Handle Problems", Configuration.CATEGORY_GENERAL, CONFIG_DEFAULT_AUTOFISH_HANDLEPROBLEMS, "[HAS SIDE EFFECTS] Re-cast when problems detected (e.g. if not in water or if a MOB has been hooked).  If enabled, non-fishing use of the fishing rod (e.g. intentionally hooking MOBs) will be affected.");
        
        if (configFile.hasChanged()) {
            configFile.save();
        }
    }

}
